integration/src/main/java/io/opencensus/interop/http/netty/OpenCensusClientInboundOutboundHandler.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright 2018, OpenCensus Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opencensus.interop.http.netty;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import io.netty.channel.ChannelDuplexHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.handler.codec.http.HttpResponse;
+import io.netty.handler.codec.http.HttpStatusClass;
+import io.netty.handler.codec.http.LastHttpContent;
+import io.opencensus.common.Scope;
+import io.opencensus.trace.Span;
+import io.opencensus.trace.SpanBuilder;
+import io.opencensus.trace.SpanContext;
+import io.opencensus.trace.Status;
+import io.opencensus.trace.Tracer;
+import io.opencensus.trace.propagation.TextFormat;
+
+/**
+ * A {@link ChannelDuplexHandler} that process both inbound and outbound messages for HTTP clients.
+ *
+ * <p>For outbound messages, it sets the propagation information into the {@link HttpRequest} and
+ * starts a new {@link Span}.
+ *
+ * <p>For inbound messages, it closes the created {@link Span} after the last message of the
+ * response is received.
+ *
+ * <p>This should be placed after the {@link io.netty.handler.codec.http.HttpClientCodec}.
+ */
+ // TODO(hailongwen): moved this into `opencensus-instrumentation-http-netty` artifact.
+public class OpenCensusClientInboundOutboundHandler extends ChannelDuplexHandler {
```

Comment:
nit: this class can be final

Reply:
Thanks! It can be made final here.
I need to re-consider how to make it extendable when moving it to another artifact, maybe using composition, not inheritance.